app/src/org/commcare/dalvik/activities/LoginActivity.java

```diff
@@ -70,12 +81,24 @@
  * @author ctsims
  */
 @ManagedUi(R.layout.screen_login)
-public class LoginActivity extends CommCareActivity<LoginActivity> implements OnItemSelectedListener {
+public class LoginActivity extends CommCareActivity<LoginActivity>
+        implements OnItemSelectedListener, RuntimePermissionRequester {
 
     private static final String TAG = LoginActivity.class.getSimpleName();
-    
+
+    private final static String[] appPermissions =

```

Comment:
Hm, are we not collecting these from something like the RuntimePermissionRequester implementing classes, for a reason? In theory we could add it as an annotation and reflect them all back out, which would prevent us from having to add them in a number of places. It would also let us collapse the logic into a unified "hasPermission()" method that grabbed the permissions from that list.